Support velocity model version 2.09

lispandfound opened the pull request:
Two changes:

1. Allow support for velocity model version 2.09. This required updating the `nzcvm` code path to account for API changes. When using the `nzcvm` code path, the code now generates the model first in HDF5 format (to allow parallel generation) and then separately converts this to EMOD3D format. The C code remains unchanged and doesn't support 2.09.
2. Relax the floating point requirements for the schema. Previously, a block in a `realisation.json` file must have a floating point value for things like depth (not int). So the following would be invalid:

``` json
{
  "dt": 1
}
```

if the `dt` variable was expected to be a `float`. Now any value that can be coerced to a floating point number via the `float` function is considered valid.

lispandfound committed:
feat(schemas): allow 2.09 in NZVM config

lispandfound committed:
feat(generate-velocity-model): use parallel nzcvm output

lispandfound committed:
use function path not module path

lispandfound committed:
fix(generate-velocity-model): copy the correct output tree in nzcvm vs nzvm cases

lispandfound committed:
fix(generate-velocity-model): do not delete HDF5 file for nzcvm case

lispandfound committed:
fix(generate-velocity-model): infer the number of cores in the nzcvm case

lispandfound committed:
feat(workflow): use CPU affinity to get allocated CPU cores robustly

lispandfound committed:
fix(utils): correct documentation for `get_available_cores`

lispandfound committed:
fix(schemas): be less zealous about floating point numbers

lispandfound committed:
fix(bb_sim): remove unused dset variable

lispandfound committed:
fix(schemas): consistently apply floating-point validation

lispandfound committed:
tests(realisation): assert numerical schema validation

lispandfound commented on workflow/scripts/generate_velocity_model.py:
```diff
@@ -204,10 +224,18 @@ def generate_velocity_model(
+            velocity_model_intermediate_path / "Velocity_Model", velocity_model_output
+        )
+
+    shutil.rmtree(velocity_model_intermediate_path)
```
`shutil.copytree` *copies* the files so this is totally wrong. But this does make me think, I could move instead of copying.

lispandfound committed:
refactor(generate_velocity_model): move VM instead of copy

lispandfound committed:
docs(generate_velocity_model): improve documentation for velocity model

lispandfound committed:
refactor(generate_velocity_model): streamline nzcvm branching statements

lispandfound committed:
fix(generate_velocity_model): remove cast

lispandfound committed:
fix: ruff fixes

lispandfound committed:
fix: remove velocity model output first

lispandfound committed:
fix: use rmtree instead of unlink

lispandfound committed:
Merge branch 'pegasus' into 2p09

joelridden commented on workflow/scripts/generate_velocity_model.py:
```diff
@@ -124,17 +124,42 @@ def run_nzvm(
-    nzcvm.generate_velocity_model(nzvm_config_ffp)
+    from velocity_modelling.constants import WriteFormat
+    from velocity_modelling.scripts import generate_3d_model
+    from velocity_modelling.tools import convert_hdf5_to_emod3d
```
Do these need to be within the function?

lispandfound replied:
There was a good reason previously, but I think it no longer applies now.

joelridden reviewed (commented):
Also need to deal with the ruff import comment

lispandfound committed:
refactor(velocity_modelling): move imports for velocity_modelling code

lispandfound committed:
fix(hf-sim): ruff error

lispandfound committed:
fix(ci): remove uv lock cache

fix(ci): fix package installation

fix(ci): dependency cache glob

explicity set pwd

fix(deptry): use env variable

add debug ls

install extras explicitly

setup virtual environment

fix uv installation in pytest

lispandfound commented:
Had to fix the CI pipeline to get the tests to pass.
